packages/fiber/tests/renderer.test.tsx

```diff
@@ -226,63 +116,155 @@ describe('renderer', () => {
-      expect(object.children.length).toBe(0)
+    expect(scene.children.length).toBe(1)
+    expect(scene.children[0]).toBeInstanceOf(THREE.Mesh)
+    // Handles geometry & material attach
+    expect((scene.children[0] as ComponentMesh).geometry).toBeInstanceOf(THREE.BoxGeometry)
+    expect((scene.children[0] as ComponentMesh).material).toBeInstanceOf(THREE.MeshStandardMaterial)
+    // Handles nested attach
+    expect(scene.children[0].userData.group).toBeInstanceOf(THREE.Group)
+    // attach bypasses scene-graph
+    expect(scene.children[0].children.length).toBe(0)
+    // attaches before presenting
+    expect(lifecycle).toStrictEqual(['attach', 'mount'])
+  })
 
-      // Detach
-      expect(object.detachedObj3d).toBeUndefined()
-      await act(async () => root.render(<hasObject3dMethods />))
-      expect(object.detachedObj3d).toBeInstanceOf(THREE.Mesh)
-    })
+  it('should handle unmount', async () => {
+    const dispose = jest.fn()
+    const childDispose = jest.fn()
+    const attachDispose = jest.fn()
+    const flagDispose = jest.fn()
+
+    const attach = jest.fn()
+    const detach = jest.fn()
+
+    const object = Object.assign(new THREE.Object3D(), { dispose: jest.fn() })
+
+    const Test = (props: JSX.IntrinsicElements['mesh']) => (
+      <mesh
+        {...props}
+        ref={(self: any) => {
+          if (!self) return
+          self.dispose = dispose
+        }}>
+        <object3D
+          ref={(self: any) => {
+            if (!self) return
+            self.dispose = childDispose
+          }}
+        />
+        <object3D
+          ref={(self: any) => {
+            if (!self) return
+            self.dispose = attachDispose
+          }}
+          attach={() => (attach(), detach)}
+        />
+        <object3D
+          dispose={null}
+          ref={(self: any) => {
+            if (!self) return
+            self.dispose = flagDispose
+          }}
+        />
+        <primitive object={object} />
+      </mesh>
+    )
+
+    const store = await act(async () => root.render(<Test />))
+    await act(async () => root.render(null))
+
+    const { scene } = store.getState()
+
+    // TODO: Scheduler isn't being flushed and Jest's mocks are clashing here.
+    // We need a way to check usage of dispose after unmount that works with Jest
+    dispose()
+    childDispose()
+    attachDispose()
+    detach()
+
+    // Cleans up scene-graph
+    expect(scene.children.length).toBe(0)
+    // Calls dispose on top-level instance
+    expect(dispose).toBeCalled()
+    // Also disposes of children
+    expect(childDispose).toBeCalled()
+    // Disposes of attached children
+    expect(attachDispose).toBeCalled()
+    // Properly detaches attached children
+    expect(attach).toBeCalledTimes(1)
+    expect(detach).toBeCalledTimes(1)
+    // Respects dispose={null}
+    expect(flagDispose).not.toBeCalled()
+    // Does not dispose of primitives
+    expect(object.dispose).not.toBeCalled()
```

Comment:
I'm not sure how much of this is scheduler callbacks not getting flushed or quirks of Jest.